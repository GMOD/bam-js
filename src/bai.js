const VirtualOffset = require('./virtualOffset')
const Chunk = require('./chunk')

const BAI_MAGIC = 21578050 // BAI\1
const MAX_BINS = 1000000

function lshift(num, bits) {
  return num << bits
}
function rshift(num, bits) {
  return num >> bits
}
/**
 * calculate the list of bins that may overlap with region [beg,end) (zero-based half-open)
 * @returns {Array[number]}
 */
function reg2bins(beg, end, minShift, depth) {
  beg -= 1 // < convert to 1-based closed
  if (beg < 1) beg = 1
  if (end > 2 ** 50) end = 2 ** 34 // 17 GiB ought to be enough for anybody
  end -= 1
  let l = 0
  let t = 0
  let s = minShift + depth * 3
  const bins = []
  for (; l <= depth; s -= 3, t += lshift(1, l * 3), l += 1) {
    const b = t + rshift(beg, s)
    const e = t + rshift(end, s)
    if (e - b + bins.length > MAX_BINS)
      throw new Error(
        `query ${beg}-${end} is too large for current binning scheme (shift ${minShift}, depth ${depth}), try a smaller query or a coarser index binning scheme`,
      )
    for (let i = b; i <= e; i += 1) bins.push(i)
  }
  return bins
}

class BAI {
  /**
   * @param {filehandle} filehandle
   * @param {function} [renameRefSeqs]
   */
  constructor({ filehandle, renameRefSeqs = n => n }) {
    this.filehandle = filehandle
    this.renameRefSeq = renameRefSeqs
  }

  _findFirstData(data, virtualOffset) {
    const currentFdl = data.firstDataLine
    if (currentFdl) {
      data.firstDataLine =
        currentFdl.compareTo(virtualOffset) > 0 ? virtualOffset : currentFdl
    } else {
      data.firstDataLine = virtualOffset
    }
  }

  async lineCount(refName) {
    const indexData = await this.parse()
    if (!indexData) return -1
    const refId = indexData.refNameToId[refName]
    const indexes = indexData.indices[refId]
    if (!indexes) return -1
    const { depth } = indexData
    const binLimit = ((1 << ((depth + 1) * 3)) - 1) / 7
    const ret = indexes.binIndex[binLimit + 1]
    return ret ? ret[ret.length - 1].minv.dataPosition : -1
  }

  /**
   * @returns {Promise} for an object like
   * `{ columnNumbers, metaChar, skipLines, refIdToName, refNameToId, coordinateType, format }`
   */
  async getMetadata() {
    const {
      columnNumbers,
      metaChar,
      format,
      coordinateType,
      skipLines,
      refIdToName,
      maxBlockSize,
      refNameToId,
      firstDataLine,
    } = await this.parse()
    return {
      columnNumbers,
      metaChar,
      format,
      coordinateType,
      skipLines,
      maxBlockSize,
      refIdToName,
      refNameToId,
      firstDataLine,
    }
  }

  // memoize
  // fetch and parse the index
  async parse() {
    const data = { bai: true, maxBlockSize: 1 << 16 }
    const bytes = await this.filehandle.readFile()

    // check TBI magic numbers
    if (bytes.readUInt32LE(0) !== BAI_MAGIC) {
      throw new Error('Not a BAI file')
    }

    data.refCount = bytes.readInt32LE(4)

    // read the indexes for each reference sequence
    data.indices = new Array(data.refCount)
    let currOffset = 8
    for (let i = 0; i < data.refCount; i += 1) {
      // the binning index
      const binCount = bytes.readInt32LE(currOffset)
      currOffset += 4
      const binIndex = {}
      for (let j = 0; j < binCount; j += 1) {
        const bin = bytes.readUInt32LE(currOffset)
        const chunkCount = bytes.readInt32LE(currOffset + 4)
        currOffset += 8
        const chunks = new Array(chunkCount)
        for (let k = 0; k < chunkCount; k += 1) {
          const u = VirtualOffset.fromBytes(bytes, currOffset)
          const v = VirtualOffset.fromBytes(bytes, currOffset + 8)
          currOffset += 16
          this._findFirstData(data, u)
          chunks[k] = new Chunk(u, v, bin)
        }
        binIndex[bin] = chunks
      }

      data.indices[i] = { binIndex }
    }

    return data
  }

  async blocksForRange(refName, beg, end) {
    if (beg < 0) beg = 0

    const indexData = await this.parse()
    if (!indexData) return []
    const refId = indexData.refNameToId[refName]
    const indexes = indexData.indices[refId]
    if (!indexes) return []

    const { binIndex } = indexes

    const bins = reg2bins(beg, end, indexData.minShift, indexData.depth)

    let l
    let numOffsets = 0
    for (let i = 0; i < bins.length; i += 1) {
      if (binIndex[bins[i]]) numOffsets += binIndex[bins[i]].length
    }

    if (numOffsets === 0) return []

    let off = []
    numOffsets = 0
    for (let i = 0; i < bins.length; i += 1) {
      const chunks = binIndex[bins[i]]
      if (chunks)
        for (let j = 0; j < chunks.length; j += 1) {
          off[numOffsets] = new Chunk(
            chunks[j].minv,
            chunks[j].maxv,
            chunks[j].bin,
          )
          numOffsets += 1
        }
    }

    if (!off.length) return []

    off = off.sort((a, b) => a.compareTo(b))

    // resolve completely contained adjacent blocks
    l = 0
    for (let i = 1; i < numOffsets; i += 1) {
      if (off[l].maxv.compareTo(off[i].maxv) < 0) {
        l += 1
        off[l].minv = off[i].minv
        off[l].maxv = off[i].maxv
      }
    }
    numOffsets = l + 1

    // resolve overlaps between adjacent blocks; this may happen due to the merge in indexing
    for (let i = 1; i < numOffsets; i += 1)
      if (off[i - 1].maxv.compareTo(off[i].minv) >= 0)
        off[i - 1].maxv = off[i].minv
    // merge adjacent blocks
    l = 0
    for (let i = 1; i < numOffsets; i += 1) {
      if (off[l].maxv.blockPosition === off[i].minv.blockPosition)
        off[l].maxv = off[i].maxv
      else {
        l += 1
        off[l].minv = off[i].minv
        off[l].maxv = off[i].maxv
      }
    }
    numOffsets = l + 1

    return off.slice(0, numOffsets)
  }
}

// this is the stupidest possible memoization, ignores arguments.
function tinyMemoize(_class, methodName) {
  const method = _class.prototype[methodName]
  if (!method)
    throw new Error(`no method ${methodName} found in class ${_class.name}`)
  const memoAttrName = `_memo_${methodName}`
  _class.prototype[methodName] = function _tinyMemoized() {
    if (!(memoAttrName in this)) this[memoAttrName] = method.call(this)
    return this[memoAttrName]
  }
}
// memoize index.parse()
tinyMemoize(BAI, 'parse')

module.exports = BAI
