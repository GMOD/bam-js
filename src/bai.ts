import Long from 'long'
import { fromBytes } from './virtualOffset'
import Chunk from './chunk'

import IndexFile from './indexFile'
import { longToNumber, optimizeChunks, BaseOpts } from './util'

const BAI_MAGIC = 21578050 // BAI\1

function roundDown(n: number, multiple: number) {
  return n - (n % multiple)
}
function roundUp(n: number, multiple: number) {
  return n - (n % multiple) + multiple
}

export default class BAI extends IndexFile {
  baiP?: Promise<Buffer>

  parsePseudoBin(bytes: Buffer, offset: number) {
    const lineCount = longToNumber(
      Long.fromBytesLE(
        Array.prototype.slice.call(bytes, offset + 16, offset + 24),
        true,
      ),
    )
    return { lineCount }
  }

  async lineCount(refId: number, opts: BaseOpts = {}) {
    const prom = await this.parse(opts)
    const index = prom.indices[refId]
    if (!index) {
      return -1
    }
    const ret = index.stats || {}
    return ret.lineCount === undefined ? -1 : ret.lineCount
  }

  fetchBai(opts: BaseOpts = {}) {
    if (!this.baiP) {
      this.baiP = this.filehandle.readFile(opts).catch(e => {
        this.baiP = undefined
        throw e
      }) as Promise<Buffer>
    }
    return this.baiP
  }

  // fetch and parse the index
  async _parse() {
    const data: { [key: string]: any } = { bai: true, maxBlockSize: 1 << 16 }
    const bytes = await this.fetchBai()

    // check BAI magic numbers
    if (bytes.readUInt32LE(0) !== BAI_MAGIC) {
      throw new Error('Not a BAI file')
    }

    data.refCount = bytes.readInt32LE(4)
    const depth = 5
    const binLimit = ((1 << ((depth + 1) * 3)) - 1) / 7

    // read the indexes for each reference sequence
    data.indices = new Array(data.refCount)
    let currOffset = 8
    for (let i = 0; i < data.refCount; i += 1) {
      // the binning index
      const binCount = bytes.readInt32LE(currOffset)
      let stats

      currOffset += 4
      const binIndex: { [key: number]: Chunk[] } = {}
      for (let j = 0; j < binCount; j += 1) {
        const bin = bytes.readUInt32LE(currOffset)
        currOffset += 4
        if (bin === binLimit + 1) {
          currOffset += 4
          stats = this.parsePseudoBin(bytes, currOffset)
          currOffset += 32
        } else if (bin > binLimit + 1) {
          throw new Error('bai index contains too many bins, please use CSI')
        } else {
          const chunkCount = bytes.readInt32LE(currOffset)
          currOffset += 4
          const chunks = new Array(chunkCount)
          for (let k = 0; k < chunkCount; k += 1) {
            const u = fromBytes(bytes, currOffset)
            const v = fromBytes(bytes, currOffset + 8)
            currOffset += 16
            this._findFirstData(data, u)
            chunks[k] = new Chunk(u, v, bin)
          }
          binIndex[bin] = chunks
        }
      }

      const linearCount = bytes.readInt32LE(currOffset)
      currOffset += 4
      // as we're going through the linear index, figure out
      // the smallest virtual offset in the indexes, which
      // tells us where the BAM header ends
      const linearIndex = new Array(linearCount)
      for (let k = 0; k < linearCount; k += 1) {
        linearIndex[k] = fromBytes(bytes, currOffset)
        currOffset += 8
        this._findFirstData(data, linearIndex[k])
      }

      data.indices[i] = { binIndex, linearIndex, stats }
    }

    return data
  }

  async indexCov(
    seqId: number,
    start?: number,
    end?: number,
    opts: BaseOpts = {},
  ): Promise<{ start: number; end: number; score: number }[]> {
    const v = 16384
    const range = start !== undefined
    const indexData = await this.parse(opts)
    const seqIdx = indexData.indices[seqId]
    if (!seqIdx) {
      return []
    }
    const { linearIndex = [], stats } = seqIdx
    if (!linearIndex.length) {
      return []
    }
    const e = end !== undefined ? roundUp(end, v) : (linearIndex.length - 1) * v
    const s = start !== undefined ? roundDown(start, v) : 0
    let depths
    if (range) {
      depths = new Array((e - s) / v)
    } else {
      depths = new Array(linearIndex.length - 1)
    }
    const totalSize = linearIndex[linearIndex.length - 1].blockPosition
    if (e > (linearIndex.length - 1) * v) {
      throw new Error('query outside of range of linear index')
    }
    let currentPos = linearIndex[s / v].blockPosition
    for (let i = s / v, j = 0; i < e / v; i++, j++) {
      depths[j] = {
        score: linearIndex[i + 1].blockPosition - currentPos,
        start: i * v,
        end: i * v + v,
      }
      currentPos = linearIndex[i + 1].blockPosition
    }
    return depths.map(d => {
      return { ...d, score: (d.score * stats.lineCount) / totalSize }
    })
  }

  /**
   * calculate the list of bins that may overlap with region [beg,end) (zero-based half-open)
   * @returns {Array[number]}
   */
  reg2bins(beg: number, end: number) {
    end -= 1
    return [
      [0, 0],
      [1 + (beg >> 26), 1 + (end >> 26)],
      [9 + (beg >> 23), 9 + (end >> 23)],
      [73 + (beg >> 20), 73 + (end >> 20)],
      [585 + (beg >> 17), 585 + (end >> 17)],
      [4681 + (beg >> 14), 4681 + (end >> 14)],
    ]
  }

  async blocksForRange(
    refId: number,
    min: number,
    max: number,
    opts: BaseOpts = {},
  ) {
    if (min < 0) {
      min = 0
    }

    const indexData = await this.parse(opts)
    if (!indexData) {
      return []
    }
    const ba = indexData.indices[refId]
    if (!ba) {
      return []
    }

    // List of bin #s that overlap min, max
    const overlappingBins = this.reg2bins(min, max)
    const chunks: Chunk[] = []

    // Find chunks in overlapping bins.  Leaf bins (< 4681) are not pruned
    for (const [start, end] of overlappingBins) {
      for (let bin = start; bin <= end; bin++) {
        if (ba.binIndex[bin]) {
          const binChunks = ba.binIndex[bin]
          for (let c = 0; c < binChunks.length; ++c) {
            chunks.push(new Chunk(binChunks[c].minv, binChunks[c].maxv, bin))
          }
        }
      }
    }

    // Use the linear index to find minimum file position of chunks that could
    // contain alignments in the region
    const nintv = ba.linearIndex.length
    let lowest = null
    const minLin = Math.min(min >> 14, nintv - 1)
    const maxLin = Math.min(max >> 14, nintv - 1)
    for (let i = minLin; i <= maxLin; ++i) {
      const vp = ba.linearIndex[i]
      if (vp) {
        if (!lowest || vp.compareTo(lowest) < 0) {
          lowest = vp
        }
      }
    }

    return optimizeChunks(chunks, lowest)
  }
}
