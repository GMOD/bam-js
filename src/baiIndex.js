const { Parser } = require('binary-parser')
const Long = require('long')

const LocalFile = require('./localFile')

class BaiIndex {
  /**
   * @param {object} args
   * @param {string} [args.path]
   * @param {string} [args.url]
   * @param {FileHandle} [args.filehandle]
   */
  constructor({ baiFilehandle, baiPath }) {
    if (baiFilehandle) {
      this.bai = baiFilehandle
    } else if (baiPath) {
      this.bai = new LocalFile(baiPath)
    }
    this.readFile = this.bai.readFile.bind(this.bai)
    this.index = this.parseIndex()
  }
  async parseIndex() {
    const data = await this.readFile()
    const parser = new Parser()
      .string('magic', { length: 4, assert: 'BAI\u0001' })
      .int32('nref')
      .array('refs', {
        length: 'nref',
        type: new Parser().int32('nbins').array('bins', {
          length: 'nbins',
          type: new Parser()
            .int32('bin')
            .int32('nchunks')
            .array('chunks', {
              length: 'nchunks',
              type: new Parser()
                .buffer('start', {
                  length: 8,
                  formatter: bytes => Long.fromBytes(bytes, true, false),
                })
                .buffer('end', {
                  length: 8,
                  formatter: bytes => Long.fromBytes(bytes, true, false),
                }),
            }),
        }),
      })
    parser.parse(data)
  }

  getIndex() {
    return this.index
  }

  /**
   * @param {number} seqId
   * @returns {Promise} true if the index contains entries for
   * the given reference sequence ID, false otherwise
   */
  async hasDataForReferenceSequence(seqId) {
    return !!(await this.index)[seqId]
  }

  /**
   * fetch index entries for the given range
   *
   * @param {number} seqId
   * @param {number} queryStart
   * @param {number} queryEnd
   *
   * @returns {Promise} promise for
   * an array of objects of the form
   * `{start, span, containerStart, sliceStart, sliceBytes }`
   */
  // async getEntriesForRange(seqId, queryStart, queryEnd) {
  //   const seqEntries = (await this.index)[seqId]
  //   if (!seqEntries) return []
  // }
}

module.exports = BaiIndex
