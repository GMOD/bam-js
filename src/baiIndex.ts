import { GenericFilehandle } from 'generic-filehandle2'

export default class BaiIndex {
  filehandle: GenericFilehandle
  constructor({ filehandle }: { filehandle: GenericFilehandle }) {
    this.filehandle = filehandle
  }
  async setup() {
    return JSON.parse(await this.filehandle.readFile('utf8'))
  }
}
