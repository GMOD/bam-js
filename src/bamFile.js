const { unzip } = require('@gmod/bgzf-filehandle')
const { CSI } = require('@gmod/tabix')
const BAI = require('./bai')
const LocalFile = require('./localFile')
const LRU = require('lru-cache')

const BAM_MAGIC = 21840194

class BamFile {
  /**
   * @param {object} args
   * @param {string} [args.bamPath]
   * @param {FileHandle} [args.bamFilehandle]
   * @param {string} [args.baiPath]
   * @param {FileHandle} [args.baiFilehandle]
   */
  constructor({
    bamFilehandle,
    bamPath,
    baiPath,
    baiFilehandle,
    csiPath,
    csiFilehandle,
  }) {
    if (bamFilehandle) {
      this.bam = bamFilehandle
    } else if (bamPath) {
      this.bam = new LocalFile(bamPath)
    }

    if (csiFilehandle) {
      this.index = new CSI({ filehandle: csiFilehandle })
    } else if (csiPath) {
      this.index = new CSI({ filehandle: new LocalFile(csiPath) })
    } else if (baiFilehandle) {
      this.index = new BAI({ filehandle: baiFilehandle })
    } else if (baiPath) {
      this.index = new BAI({ filehandle: new LocalFile(baiPath) })
    } else {
      this.index = new BAI({ filehandle: new LocalFile(`${bamPath}.bai`) })
    }
    this.options = {
      cacheSize: args.cacheSize !== undefined ? args.cacheSize : 20000,
    }

    // cache of features in a slice, keyed by the
    // slice offset. caches all of the features in a slice, or none.
    // the cache is actually used by the slice object, it's just
    // kept here at the level of the file
    this.featureCache = LRU({
      max: this.options.cacheSize,
      length: featureArray => featureArray.length,
    })
  }

  async getHeader() {
    const indexData = await this.index.parse()
    const ret = indexData.firstDataLine
      ? indexData.firstDataLine.blockPosition + 65535
      : undefined

    const buf = Buffer.allocUnsafe(ret)
    await this.bam.read(buf, 0, ret)

    const uncba = await unzip(buf)

    if (uncba.readInt32LE(0) !== BAM_MAGIC) throw new Error('Not a BAM file')
    const headLen = uncba.readInt32LE(4)

    this.header = uncba.toString('utf8', 8, 8 + headLen)
    // this.header = ''
    // for (let j = 0; j < headLen; j += 1) {
    //   this.header += String.fromCharCode(uncba[4 + j])
    // }
    return this._readRefSeqs(headLen + 8, 65535)
  }

  // the full length of the refseq block is not given in advance so this grabs a chunk and
  // doubles it if all refseqs haven't been processed
  async _readRefSeqs(start, refSeqBytes) {
    const buf = Buffer.allocUnsafe(refSeqBytes)
    await this.bam.read(buf, 0, refSeqBytes, 0)

    const uncba = await unzip(buf)
    const nRef = uncba.readInt32LE(start)
    let p = start + 4
    this.chrToIndex = {}
    this.indexToChr = []
    for (let i = 0; i < nRef; i += 1) {
      const lName = uncba.readInt32LE(p)
      const name = uncba.toString('utf8', p + 4, p + 4 + lName - 1)
      const lRef = uncba.readInt32LE(p + lName + 4)
      this.chrToIndex[name] = i
      this.indexToChr.push({ name, length: lRef })

      p = p + 8 + lName
      if (p > uncba.length) {
        console.warn(
          `BAM header is very big.  Re-fetching ${refSeqBytes} bytes.`,
        )
        return this._readRefSeqs(start, refSeqBytes * 2)
      }
    }
    return true
  }

  fetch(chr, min, max, featCallback, endCallback, errorCallback) {
    // todo regularize refseq names
    const chrId = this.chrToIndex && this.chrToIndex[chr]
    let chunks
    if (!(chrId >= 0)) {
      chunks = []
    } else {
      chunks = this.index.blocksForRange(chrId, min, max, true)
      if (!chunks) {
        errorCallback(new Errors.Fatal('Error in index fetch'))
      }
    }

    // toString function is used by the cache for making cache keys
    chunks.toString = function() {
      return this.join(', ')
    }

    try {
      this._fetchChunkFeatures(
        chunks,
        chrId,
        min,
        max,
        featCallback,
        endCallback,
        errorCallback,
      )
    } catch (e) {
      errorCallback(e)
    }
  }

  _fetchChunkFeatures(
    chunks,
    chrId,
    min,
    max,
    featCallback,
    endCallback,
    errorCallback,
  ) {
    const thisB = this

    if (!chunks.length) {
      endCallback()
      return
    }

    let chunksProcessed = 0

    // check the chunks for any that are over the size limit.  if
    // any are, don't fetch any of them
    for (let i = 0; i < chunks.length; i++) {
      const size = chunks[i].fetchedSize()
      if (size > this.chunkSizeLimit) {
        errorCallback(
          new Errors.DataOverflow(
            `Too many BAM features. BAM chunk size ${Util.commifyNumber(
              size,
            )} bytes exceeds chunkSizeLimit of ${Util.commifyNumber(
              this.chunkSizeLimit,
            )}.`,
          ),
        )
        return
      }
    }

    let haveError
    let pastStart
    array.forEach(chunks, c => {
      this.featureCache.get(c, (f, e) => {
        if (e && !haveError) errorCallback(e)
        if ((haveError = haveError || e)) {
          return
        }

        for (let i = 0; i < f.length; i++) {
          const feature = f[i]
          if (feature._refID == chrId) {
            // on the right ref seq
            if (feature.get('start') > max)
              // past end of range, can stop iterating
              break
            else if (feature.get('end') >= min)
              // must be in range
              featCallback(feature)
          }
        }
        if (++chunksProcessed == chunks.length) {
          endCallback()
        }
      })
    })
  }

  async _readChunk(chunk, callback) {
    const features = []
    // console.log('chunk '+chunk+' size ',Util.humanReadableNumber(size));
    const bufsize = chunk.fetchedSize()
    const buf = Buffer.allocUnsafe(bufsize)
    await this.data.read(buf, chunk.minv.block, bufsize)
    const data = await unzip(buf)
    this.readBamFeatures(
      new Uint8Array(data),
      chunk.minv.offset,
      features,
      callback,
    )
  }

  readBamFeatures(ba, blockStart, sink, callback) {
    const that = this
    let featureCount = 0

    const maxFeaturesWithoutYielding = 300

    while (true) {
      if (blockStart >= ba.length) {
        // if we're done, call the callback and return
        callback(sink)
        return
      } else if (featureCount <= maxFeaturesWithoutYielding) {
        // if we've read no more than 200 features this cycle, read another one
        const blockSize = readInt(ba, blockStart)
        const blockEnd = blockStart + 4 + blockSize - 1

        // only try to read the feature if we have all the bytes for it
        if (blockEnd < ba.length) {
          const feature = new BAMFeature({
            store: this.store,
            file: this,
            bytes: { byteArray: ba, start: blockStart, end: blockEnd },
          })
          sink.push(feature)
          featureCount++
        }

        blockStart = blockEnd + 1
      } else {
        // if we're not done but we've read a good chunk of
        // features, put the rest of our work into a timeout to continue
        // later, avoiding blocking any UI stuff that's going on
        window.setTimeout(() => {
          that.readBamFeatures(ba, blockStart, sink, callback)
        }, 1)
        return
      }
    }
  }
}

module.exports = BamFile
