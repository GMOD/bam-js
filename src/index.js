import BaiIndex from './baiIndex'
export { BaiIndex }
