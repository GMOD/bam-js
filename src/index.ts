import BAI from './bai'
import CSI from './csi'
import BamFile from './bamFile'
import HtsgetFile from './htsget'
import BamRecord from './record'

export { BAI, CSI, BamFile, BamRecord, HtsgetFile }
