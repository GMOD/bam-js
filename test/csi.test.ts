import { BamFile } from '../src'

class HalfAbortController {
  signal: { aborted: boolean }
  constructor() {
    this.signal = { aborted: false }
  }

  abort() {
    this.signal.aborted = true
  }
}

test('loads volvox-sorted.bam with csi index', async () => {
  const ti = new BamFile({
    bamPath: 'test/data/volvox-sorted.bam',
    csiPath: 'test/data/volvox-sorted.bam.csi',
  })
  await ti.getHeader()
  expect(ti.header).toEqual('@SQ	SN:ctgA	LN:50001\n')
  expect(ti.chrToIndex?.ctgA).toEqual(0)
  expect(ti.indexToChr?.[0]).toEqual({ refName: 'ctgA', length: 50001 })
})

test('deep check 1000 genomes', async () => {
  const ti = new BamFile({
    bamPath: 'test/data/1000genomes_hg00096_chr1.bam',
    csiPath: 'test/data/1000genomes_hg00096_chr1.bam.csi',
  })
  await ti.getHeader()
  const records = await ti.getRecordsForRange('1', 0, 1000)
  expect(records).toMatchSnapshot()
})
test('deep check 1000 genomes csi', async () => {
  const ti = new BamFile({
    bamPath: 'test/data/1000genomes_hg00096_chr1.bam',
    csiPath: 'test/data/1000genomes_hg00096_chr1.bam.csi',
  })
  await ti.getHeader()
  const records = await ti.getRecordsForRange('1', 0, 1000)
  expect(records).toMatchSnapshot()
})
test('start to deep check 1000 genomes but abort instead', async () => {
  const aborter = new HalfAbortController()
  const ti = new BamFile({
    bamPath: 'test/data/1000genomes_hg00096_chr1.bam',
    csiPath: 'test/data/1000genomes_hg00096_chr1.bam.csi',
  })
  const recordsP = ti
    .getHeader({ signal: aborter.signal as AbortSignal })
    .then(() =>
      ti.getRecordsForRange('1', 0, 1000, {
        signal: aborter.signal as AbortSignal,
      }),
    )
  aborter.abort()
  await expect(recordsP).rejects.toThrow(/aborted/)
})

test('BamFile+CSI with large coordinates', async () => {
  const b = new BamFile({
    bamPath: 'test/data/large_coords.bam',
    csiPath: 'test/data/large_coords.bam.csi',
  })
  await b.getHeader()

  const features = await b.getRecordsForRange(
    'ctgA',
    1073741824,
    1073741824 + 50000,
  )
  expect(features.length).toEqual(9596)
})

test('SAM spec pdf', async () => {
  const b = new BamFile({
    bamPath: 'test/data/samspec.bam',
    csiPath: 'test/data/samspec.bam.csi',
  })
  await b.getHeader()

  const features = await b.getRecordsForRange('ref', 1, 100)
  expect(features.length).toEqual(6)
  expect(features[2].get('sa')).toEqual('ref,29,-,6H5M,17,0;')
  expect(features[4].get('sa')).toEqual('ref,9,+,5S6M,30,1;')
})
