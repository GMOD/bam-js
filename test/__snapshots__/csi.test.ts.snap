// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`deep check 1000 genomes 1`] = `[]`;

exports[`deep check 1000 genomes csi 1`] = `[]`;
